keps/sig-cli/kustomize-secret-generator-plugins.md

```diff
@@ -0,0 +1,170 @@
+---
+title: Kustomize Secret Generator Plugins
+authors:
+  - "@sethpollack"
+owning-sig: sig-cli
+reviewers:
+  - "@monopole"
+  - "@Liujingfang1"
+approvers:
+  - "@monopole"
+  - "@Liujingfang1"
+editor: "@sethpollack"
+creation-date: 2019-02-04
+last-updated: 2019-02-04
+status: provisional
+---
+
+# Kustomize Secret Generator Plugins
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+Kustomize wants to generate Kubernetes Secret objects from general key-value pairs where the value is supposed to be a secret. Currently Kustomize only supports reading secret values from local files which raises security concerns about file lifetime and access. Reading secret values from the execution of arbitrary "commands" in a kustomization.yaml file introduces concerns in a world where kustomization files can be used directly from the internet when a user runs `kustomize build`.
+
+This proposal describes obtaining secret values safely using golang [plugins](https://golang.org/pkg/plugin/).
+
+## Motivation
+
+Not having a way to integrate `SecretGenerator` with secret management tools requires hacky workarounds.
+
+### Goals
+
+- Give users a way to intergrate `SecretGenerator` with secret management tools in a safe way.
+
+### Non-Goals
+
+- Kustomize will not handle plugin installation/management.
```

Comment:
If kustomize is going to add plugins should it try the same model as kubectl for management?

Reply:
What do you think?

Reply:
Some options for leveraging kubectl plugin model (discussed during/after sig-cli meeting):

 - Require installation of kubectl, and have kustomize execute the secret generation code as a kubectl sub-process, i.e. `exec.Command(“kubectl”, “foo”, …)` using the plugin _via_ kubectl.

 - Arrange via cli-runtime refactoring to vendor kubectl plugin code into kustomize, to track changes to argument and flag parsing and the notion of plugin subcommands. 

-  Vendor no code, and use a simplified kubctl plugin model: given a secret generation command `foo bar`, just look for a binary called `kubectl-foo` and run it, passing `bar` as an argument.  This works because the kustomize use case doesn’t have to worry about command collisions, sub-commands, listing plugins, etc.

The downside to all these options is that a plugin intended only for kustomize secret generation would appear in kubectl help, and plugin lists, etc.

Reply:
The above comments assume familiarity with how kubectl plugins work.  if interested, see https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins

Reply:
I like option 3 the best.

Reply:
I think there _are_ risks of command collisions here.  Not right now, because nobody uses kubectl plugins ;) but imagine a world where it is normal to have several unknown `kubectl-something` commands in `$PATH`.  It is highly likely that these plugins have interesting side effects when invoked, because that's the whole point of plugins.

Given that `kustomization.yaml` files will come from relatively-untrusted external sources, I would much prefer that this specific category of plugin was cleanly namespaced off from other possible kubectl plugins.  This should be as simple as invoking `kubectl-kustomize-plugin-foo ...`, rather than just `kubectl-foo`.

Reply:
Do those extensions have to be in `$PATH`? If users are not meant to invoke them directly, then they shouldn't be.

Reply:
@anguslees I don't want to use kubectl style plugins.  Was just listing the options as a thought experiment.  Your point taken though; if we did, we could avoid collisions in kubectl with more ugly namespace.   When I said the _kustomize_ use case doesn't have to worry about command collisions, i was referring to colliding with other native kustomize commands, which isn't possible since the kustomize plugin wouldn't be used as a cobra-style command.  The only trouble would be plugin-plugin collision, and we just error out if the loader detects that.

I'd never use a kustomize.yaml file from an _untrusted_ source.  Is there something in the docs that suggests such behavior is recommended?

Reply:
> I'd never use a kustomize.yaml file from an untrusted source.

I think the larger kustomize proposal is to have 3rd parties publish remote manifest repositories that may include `kustomize.yaml` (since they are in turn derived from manifests coming from another 4th party).  That means we are pulling kustomize.yaml from a sort-of-untrusted source.

I agree that actually applying the remote manifests necessarily means we are trusting the remote source to give us code/config that will be executed in our local cluster.   There are differences between what (and when) the remote source is able to execute in the cluster vs on the local machine, however.  In particular, we need to be able to do "--dry-run" -style `kubectl` operations on remote yaml without endangering the local user.